Unset CPU CFS quota when CPU sets are in use

praseodym opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Having CFS CPU quota enabled can cause performance degradation due to the (buggy) way current Linux kernels enforce this quota.

For containers that have CPU cores dedicated to them through CPU sets, any CFS CPU quota can be safely unset to prevent this issue.

This PR also aligns the implementation with the [CPU Management documentation](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/) ("CFS quota is not used to bound the CPU usage of these containers as their usage is bound by the scheduling domain itself.").

**Which issue(s) this PR fixes**:
Fixes #70585

**Does this PR introduce a user-facing change?**:
```release-note
Unset CPU CFS quota when CPU sets are in use
```

ConnorDoyle commented:
/ok-to-test

ConnorDoyle commented:
@praseodym did you collect any benchmarks showing performance for pinned containers with and without this patch? If so it would be helpful to add the results here. Thanks!

praseodym commented:
/retest

praseodym commented:
> @praseodym did you collect any benchmarks showing performance for pinned containers with and without this patch? If so it would be helpful to add the results here. Thanks!

I do not collect any benchmarks yet. Would you have any suggestions on how to execute a fair benchmark?

dims commented:
/milestone v1.15

dims commented:
/priority important-soon

dims commented:
/assign @yujuhong @derekwaynecarr

praseodym commented:
This currently also unsets CFS quota for burstable pods :/

/hold

praseodym commented:
Updated. The CPU CFS quota is now only unset when a real CPU manager policy is active and the pod is in the guaranteed class.

Unfortunately this does require moving a bit of logic from the static CPU manager policy to the generic CPU manager code, which might be considered as a layering violation. There is already some other code in the generic CPU manager code to support the static policy however.

/hold cancel

praseodym commented:
/test

praseodym commented:
/retest

derekwaynecarr commented:
can we enhance the cpu_manager_test e2e so we ensure the right outcome never regresses?

derekwaynecarr commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -309,14 +310,35 @@ func findContainerIDByName(status *v1.PodStatus, name string) (string, error) {
 }
 
-func (m *manager) updateContainerCPUSet(containerID string, cpus cpuset.CPUSet) error {
+func (m *manager) updateContainerCPUSet(pod *v1.Pod, container *v1.Container, containerID string, cpus cpuset.CPUSet) error {
```
isnt cfs quota still going to be applied at the pod cgroup with this change?

see:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/pod_container_manager_linux.go#L87
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/helpers_linux.go#L164

praseodym replied:
Yes, unfortunately it is. Running in [kind](https://kind.sigs.k8s.io/) gives me:
```
> cat /sys/fs/cgroup/cpu,cpuacct/docker/8d89c6f3cf8e5aaace1beb1adad3f38b926420ff5490d53a80a18b4457a60834/kubepods/pod71ef1b9f-8373-4dd8-9d57-a47706112750/cpu.cfs_quota_us
100000
> cat /sys/fs/cgroup/cpu,cpuacct/docker/8d89c6f3cf8e5aaace1beb1adad3f38b926420ff5490d53a80a18b4457a60834/kubepods/pod71ef1b9f-8373-4dd8-9d57-a47706112750/7f21da8aaab220ea945cca3c29c8af1217a09d2b146721b2f98b051d063c74a5/cpu.cfs_quota_us
-1
```

I checked `cpu.stat` for the container and it had `throttled_time 0`, but `cpu.stat` for the pod (one level up in the hierarchy) did have a non-zero `throttled_time` so there is some throttling going on.

Solving this is a little more complicated. Unsetting the CFS CPU quota on the pod level would always make sense if there is only one container running with static CPU policies assigned, but with multiple containers in a pod we would have to check whether that holds for all of them. Such a check could placed in the `cpu_manager`'s `reconcileState` function, but I'm not sure whether the pod resources can be easily updated. Do you know whether that is the case?

derekwaynecarr reviewed (changes requested):
please add an e2e test that validates the desired container and pod cgroup settings so we know this works and does not regress.  the pod cgroup will still have the cfs quota applied at that scope.

@ConnorDoyle can you confirm?

If it is set at the pod boundary, do we know this really has the performance benefit you were looking to achieve?

ConnorDoyle commented on pkg/kubelet/cm/cpumanager/cpu_manager.go:
```diff
@@ -309,14 +310,35 @@ func findContainerIDByName(status *v1.PodStatus, name string) (string, error) {
+
+	// When a dedicated CPU set is assigned, unset CFS CPU quota to avoid performance
+	// degradation due to the way this quota is implemented in the Linux kernel.
+	if policyName(m.policy.Name()) != PolicyNone && GuaranteedCPUs(pod, container) != 0 {
```
Not crazy about this condition... it does seem like an abstraction leak.

Could you instead add a function like `EnforceCpuQuota(pod *v1.Pod, container *v1.Container) bool` to the Policy interface and then call that from here to determine whether to override the quota?

praseodym replied:
Thanks, that's indeed a much nicer way. Done!

dims commented:
@praseodym looks like from the questions raised by @derekwaynecarr and @ConnorDoyle that this is not yet ready. so let's move the milestone to 1.16. 

/milestone v1.16

praseodym committed:
Unset CPU CFS quota when dedicated CPU set is used

Having CFS CPU quota enabled can cause performance degradation due to
the (buggy) way current Linux kernels enforce this quota.

For containers that have CPU cores dedicated to them through CPU sets,
any CFS CPU quota can be safely unset to prevent this issue.

praseodym commented:
/retest

praseodym commented:
> can we enhance the cpu_manager_test e2e so we ensure the right outcome never regresses?

Would the best way to write such a test be to follow the same route as the pods_container_manager_test e2e tests, i.e. creating a new pod to check whether `cpu.cfs_quota_us` on sysfs contains the expected value?

praseodym commented:
/retest

draveness commented:
cc/ @draveness

yguo0905 commented:
/cc @yguo0905

dims commented:
@derekwaynecarr @yujuhong Can you please review this long standing patch? Associated issue has been open since Aug 20, 2018 :(

xmudrii commented:
@derekwaynecarr @yujuhong Hello! I'm the bug triage lead for the 1.16 release cycle and considering this PR is tagged for 1.16, I'd like to check its status. The code freeze is starting on August 29th EOD PST (about 5 days from now) and once the code freeze starts, only release-blocking PRs will be considered to be merged. Is this PR still planned for the 1.16 release?

@praseodym Please note that the PR requires to be rebased before it can be merged.

praseodym commented:
@xmudrii Thanks for the heads-up! This PR needs more work, so it's not realistic for it to get merged before the 1.16 code freeze.

/milestone clear

xmudrii commented:
@praseodym Thank you for letting us know!

/milestone clear

draveness reviewed (commented):
Can we close this since it is already fixed in the kernel, or we still need this PR

https://github.com/kubernetes/kubernetes/issues/70585#issuecomment-548145135

praseodym commented:
Proper implementation of this feature (or bug workaround, rather) will get pretty complicated, so I’m fully supportive of closing this PR in favour of the kernel fix.

If anyone has any objections, please leave a comment!

dannyk81 commented:
@praseodym @chiluk @draveness although the CFS fixes should eliminate any throttling here I would still argue that enforcing CFS quotas for pods in Guaranteed QoS with cpuset (cpumanager enabled) seems wrong.

The Pods in question are isolated into dedicated cpusets where the limits of their usage is dictated by the amount of cores allocated to the cpuset.

There doesn't seem to be any logical reason to have a CFS quota applied here as well, as it will serve no purpose and only adds to confusion.

This will also solve the issue for those not able to upgrade to latest kernels where the fix is available, though I would say that this should not be be a factor indeed.

in any case, @chiluk mad respect for figuring out the fixes and working to ensure they made it into the kernel :pray: :clap:

casidiablo commented:
Since it's been more than a year since the last comment I just wanted to bump this and express my support for the previous comment. This is still an issue, even with the kernel fix.

dkulchinsky commented:
Thanks, @casidiablo! I do hope the proposed change in this PR will be reconsidered.

praseodym commented:
I no longer have any interest in working on this PR, so #70585 is probably a better place to further discuss this.
